### PR TITLE
[WFLY-14922] Initial analysis for removing the PicketLink subsystem

### PR DESCRIPTION
**Pull request by fjuma**
https://issues.redhat.com/browse/WFLY-14922
https://issues.redhat.com/browse/WFLY-15076

### CODE REVIEW

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+The Keycloak SAML Elytron adapter already exists and is already used and tested today. Testing will
+be needed to make sure no issues arise after removing the `org.picketbox` module dependency
+from the `org.keycloak.keycloak-saml-wildfly-elytron-adapter` module.
+
+== Community Documentation
+
+In addition to removing any references to PicketLink from the community documentation,
+a "migration" article will be added to indicate that the Keycloak SAML Elytron adapter
+should be used to configure SSO with SAML for WildFly. Users should refer to the Keycloak
```
Comment (OndrejKotek) [inline security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:118]: Is the Keycloak SAML Elytron adapter usable with other SAML identity providers than Keycloak? It is possible to use PicketLink SP with non-PicketLink IDPs (and users do so). In case this is not possible with EAP Keycloak SP/adapter, such fact should be clearly visible.

Comment (fjuma) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:118]: @OndrejKotek I'll double check but it does seem to be possible to use the Keycloak SAML Elytron adapter with other SAML identity providers (e.g., see https://developers.redhat.com/blog/2019/08/27/using-keycloak-instead-of-picketlink-for-saml-based-authentication).

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+adapter needs to be updated to remove the `org.picketbox` module from its dependencies (the `org.picketbox` dependency
+does not appear to be used, it just needs to be removed).
+
+=== Nice-to-Have Requirements
+
+N/A
+
+=== Non-Requirements
+
+* The Keycloak SAML Elytron adapter does not ship with WildFly. It needs to be installed separately by the
```
[the remark this answers is not in the document]
Comment (fjuma) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:97]: We could provide a Galleon layer for this but that would a follow-up RFE I think.

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+To be discussed.
+
+The Keycloak SAML Elytron adapter already exists and is already used and tested today. Testing will
+be needed to make sure no issues arise after removing the `org.picketbox` module dependency
+from the `org.keycloak.keycloak-saml-wildfly-elytron-adapter` module.
+
+== Community Documentation
+
+In addition to removing any references to PicketLink from the community documentation,
+a "migration" article will be added to indicate that the Keycloak SAML Elytron adapter
```
Comment (OndrejKotek) [inline security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:117]: Do we want to describe also a migration path from PicketLink STS to Apache CXF STS?

Comment (fjuma) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:117]: @OndrejKotek I'll look into this and get back to you.

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+* [x] Bootable jar
+
+== Requirements
+
+Since this is a removal task, this section will describe the steps that will be needed as opposed to the
+set of requirements.
+
+=== Hard Requirements
+
+* The PicketLink subsystem needs to be converted to a skeleton subsystem. (It needs to be a skeleton subsystem
```
Comment (OndrejKotek) [inline security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:79]: I'm not sure about the expected behavior of a skeleton subsystem. Can you elaborate this please? Is the detailed configuration of the subsystem preserved in the new version of the management model? Can it be updated through management CLI operations? Will there be a warning in standalone mode? And the like.

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+=== Nice-to-Have Requirements
+
+N/A
+
+=== Non-Requirements
+
+* The Keycloak SAML Elytron adapter does not ship with WildFly. It needs to be installed separately by the
+user. Thus, migrating from SSO with SAML using PicketLink to SSO with SAML via the Keycloak adapter won't be
+something that's handled by the server migration tool. The user will need to install the Keycloak adapter on
```
Comment (OndrejKotek) [inline security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:99]: I agree that this is out of scope of this RFE. However, there was a discussion that the migration via migration tool could be possible.

Comment (OndrejKotek) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:99]: Is https://issues.redhat.com/browse/WFLY-15623 part of this RFE?

Comment (fjuma) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:99]: Yes, WFLY-15623 will be part of this. I'll need to update the analysis to reflect that.

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc (hunk 1 above)
Comment (fjuma) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:138]: @OndrejKotek I looked into this and confirmed that this is possible.

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (fjuma) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc]: Yes, a Galleon feature pack will be added as part of a separate RFE. I've updated the analysis document to reflect this.

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc (hunk 3 above)
Comment (fjuma) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc:137]: Since Apache CXF STS was already an alternative for PicketLink STS, we won't be covering this in detail in the migration documentation. If we feel more details are needed, we'll need to ask the Web Services team to look into this. I've updated the document to reflect this.

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc (hunk 4 above)
Comment (fjuma) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc]: I've added more details in the document. Let me know if anything still isn't clear.

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc (hunk 5 above)
Comment (fjuma) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc]: I've added more information about migration to the document.

File: security/WFLY-14922_Remove_PicketLink_Subsystem.adoc (hunk 4 above)
Comment (OndrejKotek) [reply security/WFLY-14922_Remove_PicketLink_Subsystem.adoc]: It's clear now, thank you.
